Use PlainListenableActionFuture for CloseFuture

tbrooks8 opened the pull request:
Right now we use a custom future for the `CloseFuture` associated with a
channel. This is because we need special unwrapping logic to ensure that
exceptions from a future failure are a certain type (opposed to an
`UncategorizedException`). However, the current version is limiting
because we can only attach one listener.

This commit changes the `CloseFuture` to extend the
`PlainListenableActionFuture`. This change allows us to attach multiple
listeners.

s1monw commented on test/framework/src/main/java/org/elasticsearch/transport/nio/channel/ChannelConsumerAdaptor.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.util.function.Consumer;
+
+public class ChannelConsumerAdaptor implements ActionListener<NioChannel> {
```
can this be pkg private and final?

s1monw replied:
hmm can't we do:

```Java
NioChannel channel = ...;
Consumer<NioChannel> consumer = ...;
ActionListener<NioChannel> insteadOfAdaptor = ActionListener.wrap(consumer, e -> consumer.accept(channel));
```

instead of having this adaptor class?

tbrooks8 commented:
@s1monw 

I explored this:

```
NioChannel channel = ...;
Consumer<NioChannel> consumer = ...;
ActionListener<NioChannel> insteadOfAdaptor = ActionListener.wrap(consumer, e -> consumer.accept(channel));
```

The fact that `ActionListenter.wrap` takes a `CheckedConsumer` was causing a lot of problems for me. For example if I changed the listened to a `CheckConsumer`, that broke using it as the onError `Consumer` because that is just a `Consumer`. I added `ActionListener.wrap` that takes a single `Runnable` and runs it in case of success or failure and that fits my needs. Is that a suitable solution?

tbrooks8 commented:
Simon pointed out to me on Slack that I could pass the reference as `consumer::accept` as a `CheckedConsumer`. So I went with that approach and removed the `Runnable` work.
